src/script/script_instance.hpp

```diff
@@ -21,10 +23,25 @@ static const uint SQUIRREL_MAX_DEPTH = 25; ///< The maximum recursive depth for
 
 /** Runtime information about a script like a pointer to the squirrel vm and the current state. */
 class ScriptInstance {
+private:
+	/** The type of the data that follows in the savegame. */
+	enum SQSaveLoadType {
+		SQSL_INT             = 0x00, ///< The following data is an integer.
+		SQSL_STRING          = 0x01, ///< The following data is an string.
+		SQSL_ARRAY           = 0x02, ///< The following data is an array.
+		SQSL_TABLE           = 0x03, ///< The following data is an table.
+		SQSL_BOOL            = 0x04, ///< The following data is a boolean.
+		SQSL_NULL            = 0x05, ///< A null variable.
+		SQSL_ARRAY_TABLE_END = 0xFF, ///< Marks the end of an array or table, no data follows.
+	};
+
 public:
 	friend class ScriptObject;
 	friend class ScriptController;
 
+	typedef std::variant<SQInteger, std::string, SQBool,SQSaveLoadType> ScriptDataVariant;
```

Comment:
Are we aware that each element in `ScriptData` is `sizeof(std::string)` + overhead of variant large. So usually 40 bytes on 64 bits platforms and usually 28 bytes on 32 bit platforms. Not sure how big the impact is on memory usage, but for non-trivial savegames it might be considerable.

Reply:
`ScriptData` used memory is released when loading is done. But maybe using `std::unique_ptr<char *>` instead of 'std::string` would be better.